envelope/envelope_test.go

```diff
@@ -0,0 +1,204 @@
+)
+
+type failToWire struct {
+	// Embed an Enveloper so we only need to implement the methods we expect.
+	Enveloper
+}
+
+func (failToWire) ToWire() (wire.Value, error) {
+	return wire.Value{}, errors.New("failed")
+}
+
+func stringp(s string) *string {
+	return &s
+}
+
+func TestWrite(t *testing.T) {
+	tests := []struct {
+		e       Enveloper
+		want    []byte
+		wantErr bool
+	}{
+		{
+			e:       failToWire{},
+			wantErr: true,
+		},
+		{
+			e: keyvalue.GetValueHelper.Args((*tv.Key)(stringp("foo"))),
+			want: []byte{
+				0x80, 0x01, 0x00, 0x01, // version|type:4 = 1 | call
+				0x00, 0x00, 0x00, 0x08, // name length = 8
+				'g', 'e', 't', 'V', 'a', 'l', 'u', 'e', // "getValue"
+				0x00, 0x00, 0x04, 0xd2, // seqID:4 = 1234
+
+				// <struct>
+				0x0b,       // type:1 = string
+				0x00, 0x01, // id:2 = 1
+				0x00, 0x00, 0x00, 0x03, // length = 3
+				'f', 'o', 'o', // "foo"
+				0x00, // stop
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		buf := &bytes.Buffer{}
+		err := Write(protocol.Binary, buf, 1234, tt.e)
+		if tt.wantErr {
+			assert.Error(t, err, "%T should fail to be enveloped", tt.e)
+			continue
+		}
+		if !assert.NoError(t, err, "%T should not fail to be enveloped", tt.e) {
+			continue
+		}
+
+		assert.Equal(t, tt.want, buf.Bytes(), "%T enveloped mismatch")
+	}
+}
+
+func TestReadReply(t *testing.T) {
+	tests := []struct {
+		desc      string
+		bs        []byte
+		want      wire.Value
+		wantSeqID int32
+		wantErr   string
+	}{
+		{
+			desc:    "Invalid envelope",
+			bs:      []byte{0},
+			wantErr: "unexpected EOF",
+		},
+		{
+			desc: "Unexpected envelope type",
+			bs: []byte{
+				0x80, 0x01, 0x00, 0x01, // version|type:4 = 1 | call
+				0x00, 0x00, 0x00, 0x03, // name length = 3
+				'a', 'b', 'c', // "abc"
+				0x00, 0x00, 0x04, 0xd2, // seqID:4 = 1234
+
+				// <struct>
+				0x00, // stop
+			},
+			want:      wire.NewValueStruct(wire.Struct{}),
+			wantSeqID: 1234,
+			wantErr:   "unknown envelope",
+		},
+		{
+			desc: "Valid reply",
+			bs: []byte{
+				0x80, 0x01, 0x00, 0x02, // version|type:4 = 2 | reply
+				0x00, 0x00, 0x00, 0x03, // name length = 3
+				'a', 'b', 'c', // "abc"
+				0x00, 0x00, 0x04, 0xd2, // seqID:4 = 1234
+
+				// <struct>
+				0x00, // stop
+			},
+			want:      wire.NewValueStruct(wire.Struct{}),
+			wantSeqID: 1234,
+		},
+		{
+			desc: "Invalid exception",
+			bs: []byte{
+				0x80, 0x01, 0x00, 0x03, // version|type:4 = 3 | exception
+				0x00, 0x00, 0x00, 0x03, // name length = 3
+				'a', 'b', 'c', // "abc"
+				0x00, 0x00, 0x04, 0xd2, // seqID:4 = 1234
+
+				// <struct> (invalid)
+				0x08,       // type:1 = i32
+				0x00, 0x01, // id:2 = 1
+				0x00, 0x00, 0x00, 0x01, // value = 1
+				0x00, // stop
+			},
+			want: wire.NewValueStruct(wire.Struct{
+				Fields: []wire.Field{
+					{1, wire.NewValueI32(1)},
+				},
+			}),
+			wantSeqID: 1234,
+			// TODO: This should probably fail to decode. Right now, it's being ignored.
+			// wantErr:   "failed to decode exception",
```

Comment:
This won't fail to decode because type mismatches for fields are ignored and both these fields are optional